### PR TITLE
Menu

### PR DESCRIPTION
**Pull request by ivantjh**
### Scene as view layer
Split view layers into `scenes`. Each `scene` will be implemented using a `scene` interface. `Classic` will be one scene. `Time Attack` will be another scene. Same goes for the rest of the screens, An interface implementation allows us to use a `std::stack<Scene>` for the rendering of our game in `pongstar.cpp`.

### Menu
Menu is drawn to avoid additional loading of sprites. `FontManager` now supports duplication of itself to reduce initialization of textures to increase performance. Multiple fontManagers are needed if the same font needs to be of different scale or kerning on the same page.

Most of the code is just re-organizing stuff.  I have also move `pickupManager` to be owned by `pongstar` and passing a pointer of it to `messageManager`. Textures are now dynamically loaded and passed to their respective scenes.

### Next up
`pongstarbase.cpp` will be used as the base class to create `time_atk` and `classic` modes.

### CODE REVIEW

File: src/fontManager.cpp [hunk 1]
```diff
@@ -16,11 +16,38 @@ const char* colorToString(fontNS::FONT_COLOR c) {
...
+			font = new Font();
+			color = y.first;
+			*font = *y.second;
+
+			cfm->insert(colorFontPair(color, font));
+		}
+
+		fonts.insert(nameColorsPair(name, cfm));
+	}
+}
```
Comment (chesterhow) [inline src/fontManager.cpp:50]: I know we dont exactly use it but you forgot the destructor

File: src/fontManager.h [hunk 2]
```diff
@@ -45,11 +45,18 @@ class FontManager {
 	nameColorsMap fonts;
 	
 public:
+	FontManager();
 	FontManager(Graphics* graphics);
+	FontManager(const FontManager& fm);	// copy constructor
+	//FontManager& operator=(const FontManager& rhs) {};
```
Comment (chesterhow) [inline src/fontManager.h]: Whats this?

File: src/scene.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+public:
+	sceneNS::TYPE* getNextSceneType() { return nextSceneType; }
+	void clearNextSceneType() { nextSceneType = NULL; }
+
+	virtual ~Scene() {}
+
+	virtual void initialize() = 0;
+	virtual void update(float frameTime) = 0;
+	virtual void ai() = 0;
+	virtual void collisions() = 0;
```
Comment (chesterhow) [inline src/scene.h:24]: Why is there a need for `ai()` and `collisions()` in scene?

File: src/menu.cpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		selectedItemIndex = selectedItemIndex == 0 ? items.size() - 1 : selectedItemIndex - 1;
+	}
+
+	if (input->wasKeyPressed(DOWN_KEY)) {
+		selectedItemIndex = selectedItemIndex == items.size() - 1 ? 0 : selectedItemIndex + 1;
+	}
+
+	if (input->wasKeyPressed(ENTER_KEY)) {
+		nextSceneType = new sceneNS::TYPE;
+		nextSceneType = &items[selectedItemIndex];
```
Comment (chesterhow) [inline src/menu.cpp:46]: Why cant I find where this variable is initialised? Also is line 45 needed?

File: src/messageManager.h [hunk 5]
```diff
@@ -1,35 +1,37 @@
-/* In charge of relying messages to the correct entites or components */
+/*	In charge of relying messages to the correct entites or components
```
Comment (chesterhow) [inline src/messageManager.h]: relaying* lol

File: src/messageManager.h [hunk 6]
```diff
@@ -1,35 +1,37 @@
...
+	MessageManager(PickupManager* pm, std::vector<Entity*>* ev);
 	~MessageManager();
 
 	void push(Message* msg);
 	void dispatch(Message* msg);
 
 	void dispatchScore(Message* msg);
 	void dispatchEffect(Message* msg);
-	void dispatchPickup(Message* msg);
+	void dispatchPickup(Message *msg);
```
Comment (chesterhow) [inline src/messageManager.h]: No.

File: src/pongstar.cpp [hunk 7]
```diff
@@ -29,103 +40,56 @@ void Pongstar::initialize(HWND hwnd) {
...
-		throw(GameError(gameErrorNS::FATAL_ERROR, "Error initializing border texture"));
+	for (size_t i = 0; i < pongstarNS::initTextureVec.size(); i++) {
+		char textureLocation[1024], errorMsg[1024];
+		tm = new TextureManager();
+		texture = pongstarNS::initTextureVec[i];
 
-	if (!paddleTexture.initialize(graphics, PADDLE_IMAGE))
-		throw(GameError(gameErrorNS::FATAL_ERROR, "Error initializing paddle texture"));
+		sprintf(textureLocation, "%s%s.png", pongstarNS::TEXTURE_DIRECTORY, textureToString(texture));
+		sprintf(errorMsg, "Erorr initializing %s texture", textureToString(texture));
```
Comment (chesterhow) [inline src/pongstar.cpp]: Erorr

File: src/pongstar.h [hunk 8]
```diff
@@ -4,7 +4,9 @@
 
 #include <vector>
 #include <queue>
+#include <stack>
 #include <chrono>
```
Comment (chesterhow) [inline src/pongstar.h]: No need for this here anymore right? Queue too

File: src/pongstarBase.h [hunk 9]
```diff
@@ -0,0 +1,81 @@
+#ifndef _PONGSTARBASE_H
+#define _PONGSTARBASE_H
```
Comment (chesterhow) [inline src/pongstarBase.h:2]: Is this the classic mode? Why not name it `classic` instead of `pongstarbase`

**Review by chesterhow — CHANGES_REQUESTED** [review]

File: src/scene.h (hunk 3 above)
Comment (chesterhow) [reply src/scene.h:24]: Nvm i get it

File: src/pongstarBase.h (hunk 9 above)
Comment (ivantjh) [reply src/pongstarBase.h:2]: Not classic, this is the base for classic and time_atk

Comment (ivantjh) [reply src/pongstarBase.h:2]: Wip

File: src/fontManager.h (hunk 2 above)
Comment (ivantjh) [reply src/fontManager.h]: Constructor for copying fontManager. You pass in an already initialized fontManager and create a copy of it. The copy does not need to reinitialize all the font textures and can already be used.

Comment (ivantjh) [reply src/fontManager.h]: Going to remove to comment. The comment allows for this
`FontManager newFm = oldFm;`

File: src/menu.cpp (hunk 4 above)
Comment (ivantjh) [reply src/menu.cpp:46]: `nextSceneType` is initialized in the interface. Yeap, line 45 is needed, have to declare as a pointer of `sceneNS::Type` before assigning value.

File: src/messageManager.h (hunk 5 above)
Comment (ivantjh) [reply src/messageManager.h]: Changed

File: src/pongstar.h (hunk 8 above)
Comment (ivantjh) [reply src/pongstar.h]: `stack` need. Our game state machine is now done through stack. `queue` don't need,

Comment (chesterhow) [reply src/pongstar.h]: i meant chrono

**Review by chesterhow — APPROVED** [review]

**Pull request merged**